[PLIN-3600]: Tides v0.9.0-beta

jredh opened the pull request:
# Jira Issue URL

https://skuidify.atlassian.net/browse/PLIN-3600

# High-Level Description

## Adds
- Tides v0.9.0 beta release of the CLI
- Upgraded from go `1.15` to `1.18` for the use of Generics
- Adds error-checking for every single process (cobra `RunE` instead of `Run`)
- Removes the ability to use uninitialized flags (`cmd.Flags().GetString(flags.ArgString.Name)` instead of `ArgString`)
- Logs to files with different flags
- Four different levels of logging (standard,`--verbose`, `--trace`, `--diagnostic`)
- Show environment variables that can be set for default values
- Removes passwords from logs
- General Housekeeping
- Major refactor and repackaging of logic
- Tests tests tests

## Removes
- Removes "Set Variable" and "Delete Variable" functionality
- Removes "Watch" and "Pull" functionality
- XML sorting and organization

## On The Horizon
- Visual Command Editor and Execution
- Streaming Requests


# Changelog:

- tons

adewaal commented:
Here's my questions / comments from a first review:
- Is there a way to see fewer and / or simpler logs? It seems like even without `-v` I get all logs for everything that happens
- Is there a specific reason you renamed the `host` flag to `pliny-host`? I think that could add unnecessary confusion
- (You’re already aware of this I think) When I use `https://domain.pliny.webserver:3000` as the host `https://` gets appended to the front of it so the retrieve fails
- When I attempted a retrieve I got ​​ `{"statusCode":415,"error":"Unsupported Media Type","message":"Unsupported Media Type"}`, but the retrieve still seemed to work, sort of. I only got the `datasources` and `profiles` folders, and the datasources folder only contained PostgreSQL type datasources

jgoodlet commented:
Some of my feedback mirrors what Amy has stated, but I'll include anyway:

- I think the `-v` flag should illustrate execution flow and status rather than include all of the logging output. Perhaps we could dump all of the extraneous logging info into a separate log file and point the user to that file if they wish to see additional info.
- `https://` is being pre-pended to the host, whether it is specified on the command line or coming from an env variable.
-  I'm also receiving the `{"statusCode":415,"error":"Unsupported Media Type","message":"Unsupported Media Type"}` error when doing a retrieve. I _thought_ perhaps it was related to the files I was retrieving, but even removing them does not resolve said error.
- Any REST datasources were excluded from the `retrieve`.
- Not sure if the above `415` is completely disrupting the `retrieve`, but it is not successfully completing the `retrieve` when I specify a specific app to retrieve (thought it is pulling `datasources`).

jredh commented:
> Here's my questions / comments from a first review:
> 
> * Is there a way to see fewer and / or simpler logs? It seems like even without `-v` I get all logs for everything that happens

For sure. There were a lot of logs that I added with this. And there could be a "trace" level for the nitty gritty that only gets output to a file logger. Is that something you'd like me to go over?

> * Is there a specific reason you renamed the `host` flag to `pliny-host`? I think that could add unnecessary confusion

I renamed it because I added a `marina-host`. The argument should be backwards compatible. Let me double check this.

> * (You’re already aware of this I think) When I use `https://domain.pliny.webserver:3000` as the host `https://` gets appended to the front of it so the retrieve fails

I'll take care of this now.

> * When I attempted a retrieve I got ​​ `{"statusCode":415,"error":"Unsupported Media Type","message":"Unsupported Media Type"}`, but the retrieve still seemed to work, sort of. I only got the `datasources` and `profiles` folders, and the datasources folder only contained PostgreSQL type datasources

The 415 is an anomaly. Can you send me the request/url for me to reproduce?

jgoodlet commented:
@jredh I'm just testing against my local pliny development site regarding the `415`. There's nothing too crazy in there...a couple of PostgreSQL datasources and a single app and an associated page and datasource.

adewaal commented:
Pros:
- I'm not getting the 415 error anymore and the logs are a lot more manageable

Cons:
- If I don't use the verbose flag I only see a `INFO[0000] Starting retrieve` message and no indication that the retrieve failed or succeeded
- If the folder I'm retrieving to doesn't already exist (and I don't include the -v flag; I'm not sure why that matters) I get `error="zip: not a valid zip file"` and the target directory is empty

coffemanfp commented on pkg/metadata.go:
```diff
@@ -0,0 +1,131 @@
+func GetMetadataTypeDirNames() (types []string) {
+	metadataType := reflect.TypeOf(NlxMetadata{})
+
+	for i := 0; i < metadataType.NumField(); i++ {
```
Since here it's being treated with reflection, I suggest to store the NumField() in a new variable and reuse it.

Like:

`for i, nF := 0, metadataType.NumField(); i < nF; ...`

jredh commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,11 @@
+# Jira Issue URL
```
Yes. This is the template for **all** GitHub pull requests made under this repository.

jredh commented on pkg/metadata.go:
```diff
@@ -0,0 +1,131 @@
+	mType := reflect.TypeOf(NlxMetadata{})
+
+	var name string
+	for i := 0; i < mType.NumField(); i++ {
```
we don't use `mType.NumField()` in the body of the loop, so this is fine as-is.

jredh commented on pkg/metadata.go:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	if util.StringSliceContainsAnyKey(validMetadataNames, []string{
+		// Check for our metadata with .xml stripped
+		strings.TrimSuffix(filePath, ".xml"),
+		// Check for our metadata with .skuid.json stripped
+		strings.TrimSuffix(filePath, ".skuid.json"),
+		// Check for theme inline css
+		strings.TrimSuffix(filePath, ".inline.css"),
```
Ordering, I believe. This is refactored legacy code

adewaal commented:
I'm not sure if you were ready for re-review or not, but I do still see all of the issues from before

jredh commented:
> I'm not sure if you were ready for re-review or not, but I do still see all of the issues from before

nope, not yet. Going to take a look at the errors you provided.

jredh commented:
> How much trouble is it to configure the JSON files so that their content is nicely formatted and not all on one line? I suspect that'll make future version control of metadata payloads easier (and be backwards compatible)

lol this was a boolean. My bad.

jredh commented:
> If I attempt to retrieve a specific app, and the app doesn’t have any SQL datasources associated with it, I get a Not Found error and the app fails to retrieve

Making a test case for this.

jredh commented:
> If I attempt to retrieve a specific app, and the app doesn’t have any SQL datasources associated with it, I get a Not Found error and the app fails to retrieve

Fixed. Was retrieving warden despite there being no warden retrieval plan.

jredh commented:
> When I retrieved an app from a site (using the -a flag) I got the correct resources, but also profiles and variables, which I shouldn’t have

taking a look now.

jredh commented:
> When I retrieved an app from a site (using the -a flag) I got the correct resources, but also profiles and variables, which I shouldn’t have

I don't think I understand this. Sent you a message.

jredh commented:
> I did see the checksum error sporadically but I know you're working on it

do you have any specific sites or conditions this might happen for? It is very sporadic for me.

jredh commented:
@adewaal ^^ lots of comments for ya

jredh commented:
@adewaal Fixed the `--app-name` handling, I think. Please give it a check when you get the chance.

Added some tests; I'm going to write some integration tests to try and figure out where that issue lies for the checksum error.

adewaal commented:
@jredh
1. Thanks for changing the JSON files to be formatted nicely ✅ 
2. The Not Found error for apps that don't have SQL datasources isn't showing up anymore ✅ 
3. I do still see the profiles and variables folders when retrieving an app, specifically if the app has a SQL datasources, but I actually tested and this is an issue for the existing CLI as well so it's probably pliny?
4. It seems like the payload directory isn't getting overwritten in the way that I expect when I retrieve (and that's the current behavior) - if I have a directory where I've pulled the whole site, and I subsequently pull just a single app down to that same directory, I still see all site details when I'd expect to just see the single app details

jredh commented:
> I do still see the profiles and variables folders when retrieving an app, specifically if the app has a SQL datasources, but I actually tested and this is an issue for the existing CLI as well so it's probably pliny?

@adewaal just checked and we're not using the filter variables in Pliny, so you might be right? 85% certainty

jredh commented:
> It seems like the payload directory isn't getting overwritten in the way that I expect when I retrieve (and that's the current behavior) - if I have a directory where I've pulled the whole site, and I subsequently pull just a single app down to that same directory, I still see all site details when I'd expect to just see the single app details

So the way we handle files if they already exist is that we "json merge" into them. This was the library that Adam updated a little while ago. Basically, if the file exists, we're going to upsert into it. I'm not sure if we want to change that behavior or overwrite.

jgoodlet commented:
Responding to your previous answer for @adewaal:

> So the way we handle files if they already exist is that we "json merge" into them. This was the library that Adam updated a little while ago. Basically, if the file exists, we're going to upsert into it. I'm not sure if we want to change that behavior or overwrite.

I agree with her sentiments. I don't think this is desirable as it is a departure from previous behavior. The directory into which you are doing a retrieve should match what the most recent retrieve specified. For example, if at some point a whole-site retrieve was executed and then I later retrieve just a specific app into the same directory, then that directory should be "cleaned out" such that it contains only those resources relevant to the specific app.

adewaal commented:
- Retrieving a local site with profile dso permissions set up causes warden to die  - I'm not sure if this is related to your changes or not but I was curious if you see that behavior
- It seems like app access for profiles isn't always getting properly retrieved (it was there for one profile but the other profiles didn't even have the `appPermission` section when they should have)
- A profile’s DSO permissions are getting retrieved even if the profile doesn’t have access to the datasource itself at the top level (you can see whether the profile has access at the "top level" from the basic profile detail page, and that corresponds to the existence / not existence of a pliny record). If a profile doesn't have access at the "top level" to a datasource, its DSO permissions shouldn't be retrieved

jredh commented:
> I agree with her sentiments. I don't think this is desirable as it is a departure from previous behavior. The directory into which you are doing a retrieve should match what the most recent retrieve specified. For example, if at some point a whole-site retrieve was executed and then I later retrieve just a specific app into the same directory, then that directory should be "cleaned out" such that it contains only those resources relevant to the specific app.

All right, I'll add that back in. If the directory exists we'll delete it.

jredh commented:
@adewaal A lot of that sounds like Pliny to me. I'll try and investigate

jredh commented:
> * Retrieving a local site with profile dso permissions set up causes warden to die  - I'm not sure if this is related to your changes or not but I was curious if you see that behavior

Could you provide the dying log message of warden next time this happens? This is kind of hard for me to reproduce

jredh commented:
> It seems like app access for profiles isn't always getting properly retrieved (it was there for one profile but the other profiles didn't even have the appPermission section when they should have)

I'm checking between the two versions right now to see if there are any indications whether or not this is previous behavior. I'm thinking this may have something to do with the filter that I'm providing since when I split the metadata (on the other branch) there's a discrepancy between the dso permissions and app permissions returned.

jredh commented:
@adewaal when I compare the profiles that I get from the old version of the cli and PLIN-3600 profiles I get the same thing. I'm not sure if the profiles issue is something related to this PR.

jredh commented:
> * It seems like app access for profiles isn't always getting properly retrieved (it was there for one profile but the other profiles didn't even have the `appPermission` section when they should have)

I'm unable to reproduce

jredh commented:
> A profile’s DSO permissions are getting retrieved even if the profile doesn’t have access to the datasource itself at the top level (you can see whether the profile has access at the "top level" from the basic profile detail page, and that corresponds to the existence / not existence of a pliny record). If a profile doesn't have access at the "top level" to a datasource, its DSO permissions shouldn't be retrieved

Just checked again; there's no difference between this PR's functionality and the previous version of cli using the site that's available to me.

adewaal commented:
1. The error message I get back from the CLI is `ERRO[0005] error with Skuid Cloud Data Service request
ERRO[0005] Error Encountered During Run: Post "https://warden.pliny.webserver:3004/api/v2/metadata/retrieve": unexpected EOF  error="Post \"https://warden.pliny.webserver:3004/api/v2/metadata/retrieve\": unexpected EOF"` and this is what I get in warden:
<img width="1372" alt="Screen Shot 2022-08-01 at 5 30 22 PM" src="https://user-images.githubusercontent.com/11199996/182250255-545b10be-91a4-461d-8c25-9138b0cc7c9e.png">
 
But I DO get this error on the old version of CLI as well, so it's not your changes, though I'm surprised you aren't seeing the issue

2. Yep, my bad, I didn't expect it, but I can see the profiles getting DSO permissions when they shouldn't on the current CLI across all three environments so its not your changes

3. I looked into it some more and it appears that if a profile has any DSO permissions set on it, it doesn't get the `appPermissions` node. Because I needed to have DSO permissions, I could only test this on a test cluster site. To repro:
- Add apps to site
- Grant Standard profile access to an app
- Create a SQL datasource and import objects
- Go into the Standard profile and give it some DSO permissions
- Retrieve the site and notice that the Standard profile's `datasourcePermissions` are there but not its `appPermissions`

jredh commented:
> The error message I get back from the CLI is ERRO[0005] error with Skuid Cloud Data Service request ERRO[0005] Error Encountered During Run: Post "https://warden.pliny.webserver:3004/api/v2/metadata/retrieve": unexpected EOF error="Post \"https://warden.pliny.webserver:3004/api/v2/metadata/retrieve\": unexpected EOF" and this is what I get in warden:

Dang. Still unable to recreate this. Going to move onto the others since this also happens in the old version.

jredh commented:
>Create a SQL datasource and import objects

Every time I try and import an object on `https://a**-staging-*-us-trial.skuidsite.ink/` I get an error.

jredh commented:
I found a DS that I can use.

jredh commented:
> Retrieve the site and notice that the Standard profile's datasourcePermissions are there but not its appPermissions

Did this and ran the docker image and saw no differences.

jredh commented:
> Retrieve the site and notice that the Standard profile's datasourcePermissions are there but not its appPermissions

This seems to be the case between this version and the last. Confirming again

jredh commented:
> Retrieve the site and notice that the Standard profile's datasourcePermissions are there but not its appPermissions

Able to reproduce!

jredh commented:
@adewaal this change should have fixed it. The problem was that we were using `pathMap` over one individual call when I copied over this logic. I've since split the method into two calls. Rather than refactor it and share the `pathMap` I just moved it into a global variable private to the package so that we can keep track of what we've already written. Not the best practice, but it solves the problem.

jredh commented:
> The error message I get back from the CLI is ERRO[0005] error with Skuid Cloud Data Service request ERRO[0005] Error Encountered During Run: Post "https://warden.pliny.webserver:3004/api/v2/metadata/retrieve": unexpected EOF error="Post "[https://warden.pliny.webserver:3004/api/v2/metadata/retrieve\](https://warden.pliny.webserver:3004/api/v2/metadata/retrieve%5C)": unexpected EOF" and this is what I get in warden

Still unable to reproduce.

jredh commented:
> The error message I get back from the CLI is ERRO[0005] error with Skuid Cloud Data Service request ERRO[0005] Error Encountered During Run: Post "https://warden.pliny.webserver:3004/api/v2/metadata/retrieve": unexpected EOF error="Post \"https://warden.pliny.webserver:3004/api/v2/metadata/retrieve\": unexpected EOF" and this is what I get in warden:

Amy it doesn't look like this would occur with this branch of code since the stack trace goes through a package that has been removed from the tides version of the cli

jredh commented:
I think that needs to be removed from Warden

adewaal commented:
1. I no longer see that issue with warden so I'm not sure what was up with that ✅ 
2. I can see that the issue I saw previously with appPermissions not showing up in the profile file has been fixed ✅

jgoodlet reviewed (approved):
Let's do this!
